apex/LAMMPS_flow.py

```diff
@@ -0,0 +1,162 @@
+                                       image=self.dpmd_image_name, command=["python3"])
+
+        relaxcal = Step(
+            name="RelaxLAMMPS-Cal",
+            template=relax,
+            artifacts={"input_lammps": relaxmake.outputs.artifacts["task_paths"]},
+            parameters={"run_command": self.lammps_run_command},
+            with_param=argo_range(relaxmake.outputs.parameters["njobs"]),
+            key="LAMMPS-Cal-{{item}}",
+            executor=self.dispatcher_executor
+        )
+        self.relaxcal = relaxcal
+
+        relaxpost = Step(
+            name="Relaxpost",
+            template=PythonOPTemplate(RelaxPostLAMMPS, image=self.dpgen_image_name, command=["python3"]),
+            artifacts={"input_post": relaxcal.outputs.artifacts["output_lammps"],
+                       "input_all": relaxmake.outputs.artifacts["output"],
+                       "param": upload_artifact(self.relax_param)},
+            parameters={"path": cwd}
+        )
+        self.relaxpost = relaxpost
+
+        if self.flow_type == 'joint':
+            propsmake = Step(
+                name="Propsmake",
+                template=PythonOPTemplate(PropsMakeLAMMPS, image=self.dpgen_image_name, command=["python3"]),
+                artifacts={"input": relaxpost.outputs.artifacts["output_all"],
+                           "param": upload_artifact(self.props_param)},
+            )
+        else:
+            propsmake = Step(
+                name="Propsmake",
+                template=PythonOPTemplate(PropsMakeLAMMPS, image=self.dpgen_image_name, command=["python3"]),
+                artifacts={"input": upload_artifact(work_dir),
+                           "param": upload_artifact(self.props_param)},
+            )
+        self.propsmake = propsmake
+
+        props = PythonOPTemplate(RunLAMMPS,
+                                 slices=Slices("{{item}}", input_artifact=["input_lammps"],
+                                               output_artifact=["output_lammps"]), image=self.dpmd_image_name, command=["python3"])
+
+        propscal = Step(
+            name="PropsLAMMPS-Cal",
+            template=props,
+            artifacts={"input_lammps": propsmake.outputs.artifacts["task_paths"]},
+            parameters={"run_command": self.lammps_run_command},
+            with_param=argo_range(propsmake.outputs.parameters["njobs"]),
+            key="LAMMPS-Cal-{{item}}",
+            executor=self.dispatcher_executor
+        )
+        self.propscal = propscal
+
+        propspost = Step(
+            name="Propspost",
+            template=PythonOPTemplate(PropsPostLAMMPS, image=self.dpgen_image_name, command=["python3"]),
+            artifacts={"input_post": propscal.outputs.artifacts["output_lammps"],
+                       "input_all": propsmake.outputs.artifacts["output"],
+                       "param": upload_artifact(self.props_param)},
+            parameters={"path": cwd}
```

Comment:
It's better to set key for each step in case of reusing.

Reply:
Keys were added.

Reply:
revised